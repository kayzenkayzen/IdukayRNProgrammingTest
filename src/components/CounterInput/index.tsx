import React, { useCallback } from 'react';
import { View, Text, TouchableOpacity, StyleSheet } from 'react-native';
import { theme } from '@src/theme';
import { ICounterInput } from './index.d';

const styles = StyleSheet.create({
  container: { flexDirection: 'row', justifyContent: 'center', height: 40 },
  button: {
    width: 40,
    backgroundColor: '#000',
    justifyContent: 'center',
    alignItems: 'center',
    borderWidth: 1,
    borderRadius: 4,
  },
  buttonLabel: {
    fontFamily: theme.fonts.primary.bold,
    fontSize: 16,
  },
  leftButton: {
    borderBottomLeftRadius: 20,
    borderTopLeftRadius: 20,
  },
  rightButton: {
    borderBottomEndRadius: 20,
    borderTopEndRadius: 20,
  },
  value: {
    width: 60,
    backgroundColor: '#fff',
    justifyContent: 'center',
    alignItems: 'center',
    borderWidth: 1,
    borderColor: '#000',
    borderRadius: 4,
    marginHorizontal: 5,
  },
  valueLabel: {
    fontFamily: theme.fonts.primary.bold,
    fontSize: 16,
    color: '#000',
  },
});

const CounterInput: React.FC<ICounterInput> = React.memo(props => {
  const {
    id,
    buttonBackgroundColor,
    buttonLabelColor,
    buttonBorderColor,
    value,
    min,
    max,
    onPressMinus,
    onPressMore,
  } = props;

  const handleOnPressMinus = useCallback(() => {
    typeof min === 'undefined' || (typeof min !== 'undefined' && value > min)
      ? onPressMinus(id)
      : null;
  }, [onPressMinus, id, min, value]);

  const handleOnPressMore = useCallback(() => {
    typeof max === 'undefined' || (typeof max !== 'undefined' && value < max)
      ? onPressMore(id)
      : null;
  }, [onPressMore, id, max, value]);

  return (
    <View style={styles.container}>
      <TouchableOpacity
        testID="minus-button"
        style={[
          styles.button,
          styles.leftButton,
          {
            backgroundColor: buttonBackgroundColor,
            borderColor: buttonBorderColor,
          },
        ]}
        activeOpacity={1}
        onPress={handleOnPressMinus}
      >
        <Text
          style={[
            styles.buttonLabel,
            {
              color:
                typeof min === 'undefined' ||
                (typeof min !== 'undefined' && value > min)
                  ? buttonLabelColor
                  : buttonLabelColor + '30',
            },
          ]}
        >
          -
        </Text>
      </TouchableOpacity>

      <View
        style={[
          styles.value,
          {
            borderColor: buttonBorderColor,
          },
        ]}
      >
        <Text testID="value-label" style={[styles.valueLabel]}>
          {value}
        </Text>
      </View>

      <TouchableOpacity
        testID="more-button"
        style={[
          styles.button,
          styles.rightButton,
          {
            backgroundColor:
              typeof max === 'undefined' ||
              (typeof max !== 'undefined' && value < max)
                ? buttonBackgroundColor
                : buttonBackgroundColor + '40',
            borderColor:
              typeof max === 'undefined' ||
              (typeof max !== 'undefined' && value < max)
                ? buttonBorderColor
                : buttonBackgroundColor + '40',
          },
        ]}
        activeOpacity={0.8}
        onPress={handleOnPressMore}
      >
        <Text style={[styles.buttonLabel, { color: buttonLabelColor }]}>+</Text>
      </TouchableOpacity>
    </View>
  );
});

CounterInput.defaultProps = {
  buttonBackgroundColor: 'transparent',
  buttonBorderColor: '#000000',
  buttonLabelColor: '#000000',
};

export default CounterInput;
