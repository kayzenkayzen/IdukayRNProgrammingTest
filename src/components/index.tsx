export { default as CounterInput } from '@src/components/CounterInput';
