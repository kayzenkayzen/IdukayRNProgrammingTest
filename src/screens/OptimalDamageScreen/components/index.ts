export { default as AttackCombinationItem } from './AttackCombinationItem/AttackCombinationItem';
