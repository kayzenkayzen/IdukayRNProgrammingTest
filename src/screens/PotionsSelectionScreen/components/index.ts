export { default as PotionOptionItem } from './PotionOptionItem/PotionOptionItem';
