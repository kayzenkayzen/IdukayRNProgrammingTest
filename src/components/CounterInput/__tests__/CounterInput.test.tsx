import React from 'react';
import { fireEvent, render } from '@testing-library/react-native';
import { CounterInput } from '@src/components';

describe('Test CounterInput Component', () => {
  it('should renders correctly', () => {
    const wrapper = render(
      <CounterInput
        value={0}
        onPressMinus={() => jest.fn()}
        onPressMore={() => jest.fn()}
      />,
    ).toJSON();

    expect(wrapper).toMatchSnapshot();
  });

  it('value prop should render correctly', () => {
    const { getByTestId } = render(
      <CounterInput
        value={4}
        onPressMinus={() => jest.fn()}
        onPressMore={() => jest.fn()}
      />,
    );

    expect(getByTestId('value-label').props.children).toBe(4);
  });

  it('onPressMinus callback should be called', () => {
    const onPressMinus = jest.fn();

    const { getByTestId } = render(
      <CounterInput
        value={0}
        onPressMinus={onPressMinus}
        onPressMore={() => jest.fn()}
      />,
    );
    const minusButton = getByTestId('minus-button');

    fireEvent.press(minusButton);

    expect(onPressMinus).toHaveBeenCalledTimes(1);
  });

  it('onPressMore callback should be called', () => {
    const onPressMore = jest.fn();

    const { getByTestId } = render(
      <CounterInput
        value={0}
        onPressMinus={() => jest.fn()}
        onPressMore={onPressMore}
      />,
    );
    const moreButton = getByTestId('more-button');

    fireEvent.press(moreButton);

    expect(onPressMore).toHaveBeenCalledTimes(1);
  });

  it('should prevent setting the input value less than the value defined on min prop', () => {
    const onPressMinus = jest.fn();
    const { getByTestId } = render(
      <CounterInput
        value={0}
        min={0}
        onPressMinus={onPressMinus}
        onPressMore={() => jest.fn()}
      />,
    );
    const minusButton = getByTestId('minus-button');

    fireEvent.press(minusButton);

    expect(onPressMinus).toHaveBeenCalledTimes(0);
  });

  it('should prevent setting the input value greater than the value defined on max prop', () => {
    const onPressMore = jest.fn();
    const { getByTestId } = render(
      <CounterInput
        value={10}
        max={10}
        onPressMinus={() => jest.fn()}
        onPressMore={onPressMore}
      />,
    );
    const moreButton = getByTestId('more-button');

    fireEvent.press(moreButton);

    expect(onPressMore).toHaveBeenCalledTimes(0);
  });
});
