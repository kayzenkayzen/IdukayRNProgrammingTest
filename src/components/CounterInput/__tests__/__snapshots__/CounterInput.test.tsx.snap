// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test CounterInput Component should renders correctly 1`] = `
<View
  style={
    Object {
      "flexDirection": "row",
      "height": 40,
      "justifyContent": "center",
    }
  }
>
  <View
    accessible={true}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "transparent",
        "borderBottomLeftRadius": 20,
        "borderColor": "#000000",
        "borderRadius": 4,
        "borderTopLeftRadius": 20,
        "borderWidth": 1,
        "justifyContent": "center",
        "opacity": 1,
        "width": 40,
      }
    }
    testID="minus-button"
  >
    <Text
      style={
        Array [
          Object {
            "fontFamily": "SourceSansPro-Bold",
            "fontSize": 16,
          },
          Object {
            "color": "#000000",
          },
        ]
      }
    >
      -
    </Text>
  </View>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "backgroundColor": "#fff",
          "borderColor": "#000",
          "borderRadius": 4,
          "borderWidth": 1,
          "justifyContent": "center",
          "marginHorizontal": 5,
          "width": 60,
        },
        Object {
          "borderColor": "#000000",
        },
      ]
    }
  >
    <Text
      style={
        Array [
          Object {
            "color": "#000",
            "fontFamily": "SourceSansPro-Bold",
            "fontSize": 16,
          },
        ]
      }
      testID="value-label"
    >
      0
    </Text>
  </View>
  <View
    accessible={true}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "transparent",
        "borderBottomEndRadius": 20,
        "borderColor": "#000000",
        "borderRadius": 4,
        "borderTopEndRadius": 20,
        "borderWidth": 1,
        "justifyContent": "center",
        "opacity": 1,
        "width": 40,
      }
    }
    testID="more-button"
  >
    <Text
      style={
        Array [
          Object {
            "fontFamily": "SourceSansPro-Bold",
            "fontSize": 16,
          },
          Object {
            "color": "#000000",
          },
        ]
      }
    >
      +
    </Text>
  </View>
</View>
`;
